h5py/h5p.pyx

```diff
@@ -519,13 +535,25 @@ cdef class PropDCID(PropOCID):
         converted to match the array dtype.  If the array has nonzero
         rank, only the first element will contain the value.
         """
+        from .h5t import check_string_dtype
         cdef TypeID tid
+        cdef char * c_ptr = NULL
 
         check_numpy_write(value, -1)
+
+        # check for vlen strings
+        # create correct typeID and convert from c_str pointer to string
+        string_info = check_string_dtype(value.dtype)
+        if string_info is not None and string_info.length is None:
+            tid = py_create(value.dtype, logical=1)
+            ret = H5Pget_fill_value(self.id, tid.id, &c_ptr)
+            fill_value = c_ptr
+            value[0] = fill_value.decode(string_info.encoding, "surrogateescape")
```

Comment:
I would actually skip the decoding. We read all strings from datasets as bytes data by default, so returning the fill value as bytes is consistent.

Reply:
I've changed it to not decode, fixed the tests. Also updated the example code above.